ShaiRandom/GapShuffler.cs

```diff
@@ -0,0 +1,283 @@
+        /// <param name="items">Items to shuffle, which are copied into an array internally.</param>
+        public GapShufflerEnumerator(IEnhancedRandom rng, ReadOnlySpan<TItem> items)
+        {
+            _items = items.ToArray();
+            _size = _items.Count;
+            _rng = rng;
+
+            rng.Shuffle(_items);
+
+            _index = 0;
+            _current = default!;
+        }
+
+        /// <summary>
+        /// Creates an enumerator that implements the "gap shuffle" algorithm.  Instead of copying the items into
+        /// an array like the other constructor or the IList.GapShuffler extension method, this takes a reference to a
+        /// list and uses it without making a copy.
+        ///
+        /// This means that the list you give it will change pseudo-randomly as you advance the enumerator.  This is
+        /// useful if you do not care about the order of the collection after you are done with the enumerator and wish
+        /// to avoid a copy being made.
+        /// </summary>
+        /// <param name="rng">The RNG to use.</param>
+        /// <param name="items">The list of items to shuffle; the items will be shuffled in place, and no copy of this list is made.</param>
+        public GapShufflerEnumerator(IEnhancedRandom rng, ref IList<TItem> items)
+        {
+            _items = items;
+            _size = _items.Count;
+            _rng = rng;
+
+            rng.Shuffle(_items);
+
+            _index = 0;
+            _current = default!;
+        }
+
+        /// <inheritdoc />
+        public bool MoveNext()
+        {
+            if(_size == 1)
+            {
+                _current = _items[0];
+                return true;
+            }
+            if(_index >= _size)
+            {
+                int n = _size - 1;
+                int swapWith;
+                for (int i = n; i > 1; i--) {
+                    swapWith = _rng.NextInt(i);
+                    (_items[i - 1], _items[swapWith]) = (_items[swapWith], _items[i - 1]);
+                }
+                swapWith = 1 + _rng.NextInt(n);
+                (_items[n], _items[swapWith]) = (_items[swapWith], _items[n]);
+                _index = 0;
+            }
+            _current = _items[_index++];
+            return true;
+        }
+
+        /// <summary>
+        /// Returns this enumerator.
+        /// </summary>
+        /// <returns>This enumerator.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public GapShufflerEnumerator<TItem> GetEnumerator() => this;
+
+        // Explicitly implemented to ensure we prefer the non-boxing versions where possible
+        #region Explicit Interface Implementations
+        /// <summary>
+        /// This iterator does not support resetting.
+        /// </summary>
+        /// <exception cref="NotSupportedException"/>
+        void IEnumerator.Reset() => throw new NotSupportedException();
```

Comment:
This doesn't bother me, but is there a technical reason why Reset() can't be implemented? I can see the case being made that because Reset() doesn't take an IEnhancedRandom or its seed, it can't actually reset the GapShuffler back to its starting state. Since most `IEnumerator`s should reset completely when Reset() is called, and this can't, not implementing may the best option.

Reply:
Resetting the RNG is well beyond the scope of what the enumerator could reasonable do (or probably what it _should_ do, since it has no idea if the RNG is shared), yeah.  The biggest problem, though, isn't the RNG, it's the array itself.  I can't restore the array to the original ordering, since I have no idea what that original ordering is or how it compares to the original.

Reset isn't really used too much in most of the common use cases for enumerators, so I don't think this is too much of an issue.  In either case, I'm not sure how I could solve the problem of not knowing the original ordering of the list.